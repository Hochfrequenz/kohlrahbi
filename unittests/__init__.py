"""
This file is here, because this allows for best de-coupling of tests and application/library logic.
Further reading: https://docs.pytest.org/en/6.2.x/goodpractices.html#tests-outside-application-code
"""

from pathlib import Path

from efoli import EdifactFormat

path_to_test_edi_energy_mirror_repo: Path = Path(__file__).parent / "test-edi-energy-mirror-repo"
path_to_test_files_fv2310 = path_to_test_edi_energy_mirror_repo / "edi_energy_de" / Path("FV2310")
# list of pruefis to check against in test_current_state
current_state_pruefis = [
    "13002",
    "13014",
    "13023",
    "15002",
    "17002",
    "17010",
    "17114",
    "17129",
    "17201",
    "17208",
    "17301",
    "19002",
    "19007",
    "19016",
    "19103",
    "19120",
    "19131",
    "19204",
    "19301",
    "19302",
    "21002",
    "21010",
    "21015",
    "21030",
    "21039",
    "23001",
    "23009",
    "25001",
    "25003",
    "25004",
    "25006",
    "25009",
    "27001",
    "27003",
    "29001",
    "29002",
    "31001",
    "31003",
    "31007",
    "31011",
    "33004",
    "35001",
    "35004",
    "37000",
    "37002",
    "37005",
    "39000",
    "39002",
    "44001",
    "44005",
    "44009",
    "44016",
    "44023",
    "44040",
    "44053",
    "44103",
    "44105",
    "44112",
    "44117",
    "44138",
    "44140",
    "44148",
    "44163",
    "44180",
    "55003",
    "55009",
    "55014",
    "55035",
    "55041",
    "55062",
    "55074",
    "55075",
    "55077",
    "55081",
    "55086",
    "55093",
    "55105",
    "55115",
    "55127",
    "55147",
    "55157",
    "55182",
    "55200",
    "55213",
    "55555",
]
test_formats = [EdifactFormat.ORDCHG, EdifactFormat.ORDRSP, EdifactFormat.IFTSTA, EdifactFormat.IFTSTA]
