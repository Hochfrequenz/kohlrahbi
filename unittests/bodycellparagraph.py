from docx.shared import Length
from attr import define, field


@define(auto_attribs=True, kw_only=True)
class BodyCellParagraph:
    text: str = field()
    tabstop_positions: list[Length]
    left_indent_length: Length

    @text.validator
    def _check_text(self, attribute, value):
        """
        This validator checks if the amount of tabstop characters `\t` is equal to the amount of tabstop positions
        """
        if value.count("\t") != len(self.tabstop_positions):
            raise ValueError("The amount of tabs in the text attribute must match the amount of tabstop positions")
