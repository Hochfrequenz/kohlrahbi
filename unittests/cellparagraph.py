from typing import Optional

from attr import define, field
from docx.shared import Length  # type:ignore[import]


@define(auto_attribs=True, kw_only=True)
class CellParagraph:
    """
    This class provides collects the main parameters we use during parsing. Namely they are
    - text: Text in a cell paragraph
    - tabstop_positions: The postions for each tabstop
    - left_indent_length: The position where the text begins in the current paragraph

    In addition there is a validator which checks if the amount of given tabstop positions is equal to the given tabstop postions.
    """

    text: str = field()
    tabstop_positions: Optional[list[Length]] = field()
    left_indent_length: Length = field()

    @text.validator
    def _check_text(self, attribute, value) -> None:
        """
        If the text contains tabstop characters `\t`, the amount of tab stops must be equal to the amount of tabstop positions.
        """
        if "\t" in value:
            if self.tabstop_positions is None:
                raise ValueError("There is a tab in the text but the tabstop positions is None")
            elif value.count("\t") != len(self.tabstop_positions):
                raise ValueError("The amount of tabs in the text attribute must match the amount of tabstop positions")
