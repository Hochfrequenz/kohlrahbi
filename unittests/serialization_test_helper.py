"""

"""

from typing import TypeVar

from pydantic import BaseModel

T = TypeVar("T", bound=BaseModel)


def assert_serialization_roundtrip(serializable_object: T) -> T:
    """
    Perform a serialization roundtrip and check that the input object is the same after serialization and deserialization.

    :returns the deserialized_object
    """
    json_string = serializable_object.model_dump_json()
    assert json_string is not None

    deserialized_object = serializable_object.model_validate_json(json_data=json_string)

    assert isinstance(deserialized_object, type(serializable_object))

    assert deserialized_object == serializable_object
    return deserialized_object
