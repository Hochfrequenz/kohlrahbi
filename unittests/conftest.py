import docx
import pytest

from unittests.cellparagraph import CellParagraph


@pytest.fixture
def get_ahb_table_with_multiple_paragraphs():
    def _setup_ahb_table(body_cell_paragraphs: list[CellParagraph]):
        doc = docx.Document()
        table = doc.add_table(rows=1, cols=1)

        body_cell = table.rows[0].cells[0]

        # the cell comes with an empty paragraph which I could not delete.
        # So we insert the BodyCellParagraph attributes into the empty paragraph
        first_body_cell_paragprah: CellParagraph = body_cell_paragraphs[0]

        body_cell.paragraphs[0].text = first_body_cell_paragprah.text

        if first_body_cell_paragprah.tabstop_positions is not None:
            for tabstop_position in first_body_cell_paragprah.tabstop_positions:
                body_cell.paragraphs[0].paragraph_format.tab_stops.add_tab_stop(tabstop_position)

        body_cell.paragraphs[0].paragraph_format.left_indent = first_body_cell_paragprah.left_indent_length

        for paragraph_index, bcp in zip(range(1, len(body_cell_paragraphs[1:]) + 1), body_cell_paragraphs[1:]):
            # add paragraph with text
            body_cell.add_paragraph(text=bcp.text)

            # set tabstop positions
            if bcp.tabstop_positions is not None:
                for tabstop_position in bcp.tabstop_positions:
                    body_cell.paragraphs[paragraph_index].paragraph_format.tab_stops.add_tab_stop(tabstop_position)

            # set left indent lenght
            body_cell.paragraphs[paragraph_index].paragraph_format.left_indent = bcp.left_indent_length

        return table

    return _setup_ahb_table
