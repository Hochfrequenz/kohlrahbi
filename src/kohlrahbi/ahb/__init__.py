"""
This module contains the functions to scrape the AHBs for Pruefidentifikatoren.
"""

import fnmatch
import gc
import re
from datetime import date
from pathlib import Path
from typing import Any, Dict, Optional

import click
import docx
import tomlkit
from docx.document import Document
from docx.table import Table
from maus.edifact import EdifactFormatVersion

from kohlrahbi.ahbtable.ahbtable import AhbTable
from kohlrahbi.docxfilefinder import DocxFileFinder
from kohlrahbi.enums.ahbexportfileformat import AhbExportFileFormat
from kohlrahbi.logger import logger
from kohlrahbi.read_functions import (
    get_ahb_table,
    get_all_paragraphs_and_tables,
    table_header_starts_with_text_edifact_struktur,
)
from kohlrahbi.seed import Seed
from kohlrahbi.unfoldedahb import UnfoldedAhb

_pruefi_pattern = re.compile(r"^[1-9]\d{4}$")


def load_pruefi_docx_file_map_from_file(path_to_pruefi_docx_file_map_file: Path) -> dict[str, dict[str, str]]:
    """
    Loads the file which contains all known Prüfidentifikatoren.
    The file may be manually updated with the script `collect_pruefis.py`.
    """
    assert path_to_pruefi_docx_file_map_file.exists(), f"The file {path_to_pruefi_docx_file_map_file} does not exist."

    with open(path_to_pruefi_docx_file_map_file, "rb") as file:
        pruefi_docx_file_map: dict[str, dict[str, str]] = tomlkit.load(file)

    return pruefi_docx_file_map


def process_ahb_table(
    ahb_table: AhbTable,
    pruefi: str,
    output_path: Path,
    file_type: tuple[AhbExportFileFormat, ...],
) -> None:
    """
    Process the ahb table.
    """
    unfolded_ahb = UnfoldedAhb.from_ahb_table(ahb_table=ahb_table, pruefi=pruefi)

    if AhbExportFileFormat.XLSX in file_type:
        unfolded_ahb.dump_xlsx(output_path)
    if AhbExportFileFormat.FLATAHB in file_type:
        unfolded_ahb.dump_flatahb_json(output_path)
    if AhbExportFileFormat.CSV in file_type:
        unfolded_ahb.dump_csv(output_path)
    del unfolded_ahb


# pylint:disable=anomalous-backslash-in-string
def get_valid_pruefis(list_of_pruefis: list[str], all_known_pruefis: Optional[list[str]] = None) -> list[str]:
    """
    This function returns a list with only those pruefis which match the pruefi_pattern "^[1-9]\\d{4}$".
    It also supports unix wildcards like '*' and '?' if a list of known pruefis is given.
    E.g. '11*' for all pruefis starting with '11' or '*01' for all pruefis ending with '01'.
    """
    result: set[str] = set()

    for pruefi in list_of_pruefis:
        if ("*" in pruefi or "?" in pruefi) and all_known_pruefis:
            filtered_pruefis = fnmatch.filter(all_known_pruefis, pruefi)
            result = result.union(filtered_pruefis)
        elif _pruefi_pattern.match(pruefi):
            result.add(pruefi)

    return sorted(list(result))


def validate_pruefis(pruefis: list[str]) -> list[str]:
    """
    Validate the pruefi_to_file_mapping parameter.
    """
    valid_pruefis = get_valid_pruefis(pruefis)
    if not valid_pruefis:
        click.secho("⚠️ There are no valid pruefidentifkatoren.", fg="red")
        raise click.Abort()
    return valid_pruefis


def process_pruefi(
    pruefi: str,
    path_to_ahb_docx_file: Path,
    output_path: Path,
    file_type: tuple[AhbExportFileFormat, ...],
) -> None:
    """
    Process one pruefi.
    If the input path ends with .docx, we assume that the file containing the pruefi is given.
    Therefore we only access that file.
    """

    # TODO try to cache document objects cause it is slow to read them from disk # pylint: disable=fixme
    doc = docx.Document(str(path_to_ahb_docx_file))

    if not doc:
        return

    ahb_table = get_ahb_table(document=doc, pruefi=pruefi)
    if not ahb_table:
        return

    process_ahb_table(ahb_table, pruefi, output_path, file_type)
    del ahb_table.table
    del ahb_table
    del doc
    gc.collect()


def get_ahb_documents_path(base_path: Path, version: str) -> Path:
    """Returns the path to the AHB documents for the specified format version."""
    path = base_path / f"edi_energy_de/{version}"
    if not path.exists():
        raise FileNotFoundError(f"The specified path {path.absolute()} does not exist.")
    return path


def find_pruefidentifikatoren(path: Path) -> Dict[str, str]:
    """finds pruefis in given dir"""
    pruefis = {}

    ahb_file_finder = DocxFileFinder.from_input_path(input_path=path)
    ahb_file_finder.filter_for_latest_ahb_docx_files()

    for docx_path in ahb_file_finder.paths_to_docx_files:
        pruefis.update(extract_pruefis_from_docx(docx_path))
    return dict(sorted(pruefis.items()))


def extract_pruefis_from_docx(docx_path: Path) -> Dict[str, str]:
    """Extracts the Prüfidentifikatoren from the given docx file."""
    doc = docx.Document(str(docx_path))
    pruefis: dict[str, str] = {}
    for item in get_all_paragraphs_and_tables(doc):
        if (
            isinstance(item, Table)
            and table_header_starts_with_text_edifact_struktur(item)
            and table_header_contains_text_pruefidentifikator(item)
        ):
            # pylint:disable=not-an-iterable
            pruefis.update({pruefi: docx_path.name for pruefi in extract_pruefis_from_table(item)})
    return pruefis


def save_pruefi_map_to_toml(pruefis: Dict[str, str], version: str) -> None:
    """Saves the pruefis to file mapping to a toml file."""
    output_filename = f"{version}_pruefi_docx_filename_map.toml"
    output_file_path = Path(__file__).parents[1] / "cache" / output_filename
    output_file_path.parent.mkdir(parents=True, exist_ok=True)

    toml_data = {"meta_data": {"updated_on": date.today()}, "pruefidentifikatoren": pruefis}
    with open(output_file_path, "w", encoding="utf-8") as f:
        tomlkit.dump(toml_data, f)
        logger.info("🎉 Successfully updated %s and saved it at %s.", output_filename, output_file_path)


def log_no_pruefis_warning(version: str, path: Path) -> None:
    """Logs a warning if no Prüfidentifikatoren were found."""
    logger.warning("No Prüfidentifikatoren found in the AHB documents for format version %s at %s.", version, path)


def get_default_pruefi_map(path: Path) -> Dict[str, str]:
    """Returns a default pruefi map if no pruefis were found."""
    return {"⚠️ No Prüfidentifikatoren found": f"No AHB documents found in {path}."}


def extract_pruefis_from_table(table: Table) -> list[str]:
    """Extracts the Prüfidentifikatoren from given table."""
    seed = Seed.from_table(docx_table=table)
    logger.info("Found a table with the following pruefis: %s", seed.pruefidentifikatoren)
    return seed.pruefidentifikatoren


def table_header_contains_text_pruefidentifikator(table: Table) -> bool:
    """Checks if the table header contains the text 'Prüfidentifikator'."""
    return table.row_cells(0)[-1].paragraphs[-1].text.startswith("Prüfidentifikator")  # type:ignore[no-any-return]


def get_pruefi_to_file_mapping(basic_input_path: Path, format_version: EdifactFormatVersion) -> dict[str, str]:
    """Returns the pruefi to file mapping. If the cache file does not exist, it creates it."""
    default_path_to_cache_file = Path(__file__).parents[1] / "cache" / f"{format_version}_pruefi_docx_filename_map.toml"

    if default_path_to_cache_file.exists():
        pruefi_to_file_mapping_cache = load_pruefi_docx_file_map_from_file(default_path_to_cache_file)
        pruefi_to_file_mapping = pruefi_to_file_mapping_cache.get("pruefidentifikatoren")
        if isinstance(pruefi_to_file_mapping, type(None)):
            raise ReferenceError(f"Could not find pruefidentifikatoren in {default_path_to_cache_file}")
        return dict(pruefi_to_file_mapping)

    path_to_docx_files = basic_input_path / Path(f"edi_energy_de/{format_version}")
    pruefi_to_file_mapping = find_pruefidentifikatoren(path_to_docx_files)
    save_pruefi_map_to_toml(pruefi_to_file_mapping, format_version.value)
    return pruefi_to_file_mapping


def reduce_pruefi_to_file_mapping(pruefi_to_file_mapping: dict[str, str], pruefis: list[str]) -> dict[str, str]:
    """
    If the user provided pruefis, we filter the pruefi_to_file_mapping for these pruefis.
    """
    return {pruefi: filename for pruefi, filename in pruefi_to_file_mapping.items() if pruefi in pruefis}


def scrape_pruefis(
    pruefis: list[str],
    basic_input_path: Path,
    output_path: Path,
    file_type: tuple[AhbExportFileFormat, ...],
    format_version: EdifactFormatVersion,
) -> None:
    """
    starts the scraping process for provided pruefi_to_file_mappings
    """
    pruefi_to_file_mapping = get_pruefi_to_file_mapping(
        basic_input_path=basic_input_path, format_version=format_version
    )
    if len(pruefis) > 0:
        validated_pruefis = validate_pruefis(pruefis)
        pruefi_to_file_mapping = reduce_pruefi_to_file_mapping(pruefi_to_file_mapping, validated_pruefis)

    for pruefi, filename in pruefi_to_file_mapping.items():
        try:
            logger.info("start looking for pruefi '%s'", pruefi)
            if not filename:
                logger.warning("No filename for pruefi '%s' provided", pruefi)
                continue
            path_to_ahb_docx_file = (
                basic_input_path / Path("edi_energy_de") / Path(format_version.name) / Path(filename)
            )
            process_pruefi(pruefi, path_to_ahb_docx_file, output_path, file_type)
        except FileNotFoundError:
            logger.exception("File not found for pruefi '%s'", pruefi)
        # sorry for the pokemon catch
        except Exception as e:  # pylint: disable=broad-except
            logger.exception("Error processing pruefi '%s': %s", pruefi, str(e))
