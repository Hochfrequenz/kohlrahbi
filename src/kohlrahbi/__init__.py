"""
kohlrahbi is a package to scrape AHBs (in docx format)
"""
import gc
import re
import sys
from pathlib import Path
from typing import Any

import click
import docx  # type:ignore[import]
import tomlkit

from kohlrahbi.ahb.ahbtable import AhbTable
from kohlrahbi.ahbfilefinder import AhbFileFinder
from kohlrahbi.logger import logger
from kohlrahbi.read_functions import get_ahb_table
from kohlrahbi.unfoldedahb.unfoldedahbtable import UnfoldedAhb

_pruefi_pattern = re.compile(r"^[1-9]\d{4}$")


def get_valid_pruefis(list_of_pruefis: list[str]) -> list[str]:
    """
    This function returns a new list with only those pruefis which match the pruefi_pattern.
    """
    valid_pruefis: list[str] = [pruefi for pruefi in list_of_pruefis if _pruefi_pattern.match(pruefi)]
    return valid_pruefis


def check_python_version():
    """
    Check if the Python interpreter is greater or equal to 3.11
    """
    if sys.version_info.major != 3 or sys.version_info.minor < 11:
        raise click.Abort(
            f"""Python >=3.11 is required to run this script but you use Python
{sys.version_info.major}.{sys.version_info.minor}"""
        )


def check_output_path(path: Path) -> None:
    """
    Checks if the given path exists and if not it asks the user if they want to create the given directory.
    """
    if not path.exists():
        click.secho("‚ö†Ô∏è The output directory does not exist.", fg="red")

        if click.confirm(f"Should I try to create the directory at '{path}'?", default=True):
            try:
                path.mkdir(exist_ok=True)
                click.secho(f"üìÇ The output directory is created at {path.absolute()}.", fg="yellow")
            except FileNotFoundError as fnfe:
                click.secho(
                    "üò± There was an path error. I can only create a new directory in an already existing directory.",
                    fg="red",
                )
                click.secho(f"Your given path is '{path}'", fg="red")
                click.secho(str(fnfe), fg="red")
                raise click.Abort()

        else:
            click.secho("üëã Alright I will end this program now. Have a nice day.", fg="green")
            raise click.Abort()


def load_all_known_pruefis_from_file(
    path_to_all_known_pruefis: Path = Path(__file__).parent / Path("all_known_pruefis.toml"),
) -> list[str]:
    """
    Loads the file which contains all known Pr√ºfidentifikatoren.
    The file may be manually updated with the script `collect_pruefis.py`.
    """

    with open(path_to_all_known_pruefis, "rb") as file:
        state_of_kohlrahbi: dict[str, Any] = tomlkit.load(file)

    meta_data_section = state_of_kohlrahbi.get("meta_data")
    content_section = state_of_kohlrahbi.get("content")

    if meta_data_section is None:
        click.secho(f"There is no 'meta_data' section in the provided toml file: {path_to_all_known_pruefis}", fg="red")
        raise click.Abort()
    if content_section is None:
        click.secho(f"There is no 'content' section in the toml file: {path_to_all_known_pruefis}", fg="red")
        raise click.Abort()

    pruefis: list[str] = content_section.get("pruefidentifikatoren")
    return pruefis


@click.command()
@click.option(
    "-p",
    "--pruefis",
    default=[],
    required=False,
    help="Five digit number like 11042.",
    multiple=True,
)
@click.option(
    "-i",
    "--input_path",
    type=click.Path(exists=True, dir_okay=True, file_okay=False, path_type=Path),
    prompt="Input directory",
    help="Define the path to the folder with the docx AHBs.",
)
@click.option(
    "-o",
    "--output_path",
    type=click.Path(exists=False, dir_okay=True, file_okay=False, path_type=Path),
    default="output",
    prompt="Output directory",
    help="Define the path where you want to save the generated files.",
)
@click.option(
    "--file-type",
    type=click.Choice(["flatahb", "csv", "xlsx"], case_sensitive=False),
    multiple=True,
)
@click.option(
    "--assume-yes",
    "-y",
    is_flag=True,
    help="Confirm all prompts automatically.",
)
# pylint: disable=too-many-branches, too-many-statements, too-many-locals
def main(pruefis: list[str], input_path: Path, output_path: Path, file_type: list[str], assume_yes: bool):
    """
    A program to get a machine readable version of the AHBs docx files published by edi@energy.
    """
    check_python_version()

    if not assume_yes:
        check_output_path(path=output_path)
    else:
        if output_path.exists():
            click.secho(f"The output directory '{output_path}' exists already.", fg="yellow")
        else:
            output_path.mkdir(parents=True)
            click.secho(f"I created a new directory at {output_path}", fg="yellow")

    if not any(pruefis):
        click.secho("‚òùÔ∏è No pruefis were given. I will parse all known pruefis.", fg="yellow")
        pruefis = load_all_known_pruefis_from_file()
    if not any(file_type):
        message = "‚Ñπ You did not provide any value for the parameter --file-type. No files will be created."
        click.secho(message, fg="yellow")
        logger.warning(message)

    valid_pruefis: list[str] = get_valid_pruefis(list_of_pruefis=pruefis)
    if not any(valid_pruefis):
        click.secho("‚ö†Ô∏è There are no valid pruefidentifkatoren.", fg="red")
        raise click.Abort()

    if len(valid_pruefis) != len(pruefis):
        click.secho("‚òùÔ∏è Not all given pruefidentifikatoren are valid.", fg="yellow")
        click.secho(f"I will continue with the following valid pruefis: {valid_pruefis}.", fg="yellow")
    path_to_document_mapping: dict[Path, docx.Document] = {}

    for pruefi in valid_pruefis:
        try:
            logger.info("start looking for pruefi '%s'", pruefi)
            ahb_file_finder = AhbFileFinder.from_input_path(input_path=input_path)
            ahb_file_paths: list[Path] = ahb_file_finder.get_docx_files_which_may_contain_searched_pruefi(
                searched_pruefi=pruefi
            )

            if not any(ahb_file_paths):
                logger.warning("No docx file was found for pruefi '%s'", pruefi)
                continue

            for ahb_file_path in ahb_file_paths:
                if not (doc := path_to_document_mapping.get(ahb_file_path, None)):
                    try:
                        doc = docx.Document(ahb_file_path)  # Creating word reader object.
                        path_to_document_mapping[ahb_file_path] = doc
                        logger.debug("Saved %s document in cache", ahb_file_path)  # to not re-read it every time
                    except IOError as ioe:
                        logger.exception("There was an error opening the file '%s'", ahb_file_path, exc_info=True)
                        raise click.Abort() from ioe

                logger.info("start reading docx file '%s'", str(ahb_file_path))

                ahb_table: AhbTable | None = get_ahb_table(
                    document=doc,
                    pruefi=pruefi,
                )

                if ahb_table is None:
                    continue

                if isinstance(ahb_table, AhbTable):
                    unfolded_ahb = UnfoldedAhb.from_ahb_table(ahb_table=ahb_table, pruefi=pruefi)

                    if "xlsx" in file_type:
                        logger.info("üíæ Saving xlsx file %s", pruefi)
                        unfolded_ahb.dump_xlsx(path_to_output_directory=output_path)

                    if "flatahb" in file_type:
                        logger.info("üíæ Saving flatahb file %s", pruefi)
                        unfolded_ahb.dump_flatahb_json(output_directory_path=output_path)

                    if "csv" in file_type:
                        logger.info("üíæ Saving csv file %s", pruefi)
                        unfolded_ahb.dump_csv(path_to_output_directory=output_path)
                    break
        except Exception as general_error:  # pylint:disable=broad-except
            logger.exception(
                "There was an uncaught error while processing the pruefi '%s': %s",
                pruefi,
                str(general_error),
                exc_info=True,
            )
            continue
        del ahb_table
        if "unfolded_ahb" in locals():
            del unfolded_ahb
        gc.collect()


if __name__ == "__main__":
    # the parameter arguments gets provided over the CLI
    main()  # pylint:disable=no-value-for-parameter
