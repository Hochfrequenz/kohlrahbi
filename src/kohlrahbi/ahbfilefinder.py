"""
This module contains the AhbFileFinder class.
"""
from itertools import groupby
from pathlib import Path

import attrs
from maus.edifact import EdifactFormat, get_format_of_pruefidentifikator

from kohlrahbi.logger import logger


@attrs.define(auto_attribs=True, kw_only=True)
class AhbFileFinder:
    """
    This class is responsible for finding the AHB files in the input directory.
    """

    paths_to_docx_files: list[Path]

    @classmethod
    def from_input_path(cls, input_path: Path) -> "AhbFileFinder":
        """
        Create an AhbFileFinder object from the input path.
        """

        ahb_file_paths: list[Path] = [
            path for path in input_path.iterdir() if path.is_file() if path.suffix == ".docx" if "AHB" in path.name
        ]
        if not any(ahb_file_paths):  # this is suspicious at least
            logger.warning("The directory '%s' does not contain any AHB docx files.", input_path.absolute())
        return cls(paths_to_docx_files=ahb_file_paths)

    @staticmethod
    def get_first_part_of_ahb_docx_file_name(path_to_ahb_document: Path) -> str:
        """
        Return the first part of the AHB docx file name.
        The first part contains the information about the EDIFACT formats.
        """

        return path_to_ahb_document.name.split("-")[0]

    def filter_for_latest_ahb_docx_files(self) -> None:
        """
        Filter the list of AHB docx paths for the latest AHB docx files.
        The latest files contain `LesefassungmitFehlerkorrekturen` in their file names.
        """
        result: list[Path] = []

        groups: dict[str, list[Path]] = {}  # the key is the first part of the file name, the values are matching files
        logger.debug("The list self.paths_to_docx_files contains %i entries", len(self.paths_to_docx_files))
        for key, group in groupby(
            sorted(self.paths_to_docx_files, key=AhbFileFinder.get_first_part_of_ahb_docx_file_name),
            AhbFileFinder.get_first_part_of_ahb_docx_file_name,
        ):
            groups[key] = list(group)

        for group_items in groups.values():
            if len(group_items) == 1:
                result.append(group_items[0])
            else:
                for path in group_items:
                    error_corrected_version_is_present = (
                        "KonsolidierteLesefassungmitFehlerkorrekturen" in path.name
                        or "AußerordentlicheVeröffentlichung" in path.name
                    )
                    if error_corrected_version_is_present:
                        result.append(path)

        self.paths_to_docx_files = result

    def filter_docx_files_for_edifact_format(self, edifact_format: EdifactFormat) -> None:
        """
        Returns a list of docx files which contain the given edifact format.
        """

        self.paths_to_docx_files = [path for path in self.paths_to_docx_files if str(edifact_format) in path.name]

    def get_docx_files_which_may_contain_searched_pruefi(self, searched_pruefi: str) -> list[Path]:
        """
        This functions takes a pruefidentifikator and returns a list of docx files which can contain the searched pruefi
        Unfortunately, it is not clear in which docx the pruefidentifikator you are looking for is located.
        A 11042 belongs to the UTILMD format. However, there are seven docx files that describe the UTILMD format.
        A further reduction of the number of files is not possible with the pruefidentifikator only.
        """

        edifact_format = get_format_of_pruefidentifikator(searched_pruefi)
        if edifact_format is None:
            logger.exception("❌ There is no known format for the prüfi '%s'.", searched_pruefi)
            raise ValueError(f"There is no known format for the prüfi '{searched_pruefi}'.")

        self.filter_for_latest_ahb_docx_files()
        self.filter_docx_files_for_edifact_format(edifact_format=edifact_format)
        if (
            edifact_format == EdifactFormat.UTILMD
            and searched_pruefi.startswith("11")
            and all("202310" in path.name for path in self.paths_to_docx_files)
        ):
            logger.info(
                # pylint:disable=line-too-long
                "You searched for a UTILMD prüfi %s starting with the soon deprecated prefix '11' but all relevant files %s are valid from 2023-10 onwards. They won't contain any match.",
                searched_pruefi,
                ", ".join([path.name for path in self.paths_to_docx_files]),
            )
            return []
        return self.paths_to_docx_files
