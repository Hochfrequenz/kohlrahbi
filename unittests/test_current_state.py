"""
This test will check if we get the same results during the refactoring process.
"""

import logging
from itertools import groupby
from pathlib import Path
from typing import Union

import pandas as pd
import pytest
from click.testing import CliRunner, Result

from kohlrahbi import cli
from unittests import path_to_test_edi_energy_mirror_repo, path_to_test_files_fv2310

# Setup basic configuration for logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")


runner: CliRunner = CliRunner()


def get_csv_paths(root_dir) -> list[Path]:
    """Walk through root_dir and return paths of all CSV files."""
    root_path = Path(root_dir)
    return [path for path in root_path.rglob("*.csv")]


def are_csv_files_equal(actual_file, expected_file) -> bool:
    """Read two CSV files into pandas DataFrames and compare them."""
    actual_df = pd.read_csv(actual_file)
    expected_df = pd.read_csv(expected_file)
    return actual_df.equals(expected_df)


def compare_csv_files(actual_output_dir: Path, expected_output_dir: Path):
    """
    Compare CSV files in expected-output and actual-output directories.
    """

    # Get paths of all CSV files in both directories
    path_to_actual_csv_files = get_csv_paths(actual_output_dir)
    path_to_expected_csv_files = get_csv_paths(expected_output_dir)

    # combine the paths of the actual and expected files
    combined_paths = path_to_actual_csv_files + path_to_expected_csv_files

    # group the paths by their filename
    combined_paths.sort(key=lambda x: x.name)
    grouped_paths = groupby(combined_paths, key=lambda x: x.name)

    differences: dict[str, list[Path] | Path] = {}

    for csv_file_name, paths in grouped_paths:
        paths_as_list = list(paths)
        if len(paths_as_list) == 2:
            if not are_csv_files_equal(paths_as_list[0], paths_as_list[1]):
                differences[csv_file_name] = [paths_as_list[0], paths_as_list[1]]
        else:
            differences[csv_file_name] = paths_as_list[0]

    # Report differences
    if differences:
        for csv_file_name, paths_to_csv_files in differences.items():
            if isinstance(paths_to_csv_files, list):
                logging.error("Files %s and %s are different.", paths_to_csv_files[0], paths_to_csv_files[1])
            else:
                logging.error(
                    "The file '%s' does not have a corresponding comparison file in the expected directory.",
                    paths_to_csv_files,
                )
        assert False
    else:
        logging.info("All files are the same.")
        assert True


class TestCli:
    """
    This class contains the unit tests for the CLI tool.
    """

    @pytest.mark.parametrize(
        "argument_options, expected_response",
        [
            pytest.param(
                [
                    "ahb",
                    "--format-version",
                    "FV2310",
                    "--file-type",
                    "csv",
                ],
                {"exit_code": 0, "output_snippet": ""},
                id="proof of concept",
            )
        ],
    )
    def test_kohlrahbi_cli_with_valid_arguments(
        self, argument_options: list[str], expected_response: dict[str, Union[str, int]], snapshot
    ):
        """
        This test runs the CLI tool with valid arguments and checks if the output is as expected.
        """

        actual_output_dir = path_to_test_files_fv2310 / "actual-output"
        expected_output_dir = path_to_test_files_fv2310 / "expected-output"

        argument_options.extend(
            [
                "--edi-energy-mirror-path",
                str(path_to_test_edi_energy_mirror_repo),
                "--output-path",
                str(actual_output_dir),
            ]
        )

        # Call the CLI tool with the desired arguments
        response: Result = runner.invoke(cli, argument_options)

        assert response.exit_code == expected_response.get("exit_code")
        expected_output_snippet = expected_response.get("output_snippet")
        if expected_output_snippet is not None and isinstance(expected_output_snippet, str):
            assert expected_output_snippet in response.output
        else:
            assert False  # break the test if the output_snippet is None

        # Check if the generated files are the same as the expected files
        for file in actual_output_dir.iterdir():
            assert file == snapshot
        # compare_csv_files(
        #     actual_output_dir=actual_output_dir,
        #     expected_output_dir=expected_output_dir,
        # )
