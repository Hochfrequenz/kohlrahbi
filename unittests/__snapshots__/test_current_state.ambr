# serializer version: 1
# name: TestCli.test_kohlrahbi_cli_with_valid_arguments[proof of concept]
  WindowsPath('C:/GitRepos/kohlrahbi/unittests/test-edi-energy-mirror-repo/edi_energy_de/FV2310/actual-output/.gitkeep')
# ---
# name: TestCli.test_kohlrahbi_cli_with_valid_arguments[proof of concept].1
  WindowsPath('C:/GitRepos/kohlrahbi/unittests/test-edi-energy-mirror-repo/edi_energy_de/FV2310/actual-output/2024-03-30_change_histories.xlsx')
# ---
# name: TestCli.test_kohlrahbi_cli_with_valid_arguments[proof of concept].2
  WindowsPath('C:/GitRepos/kohlrahbi/unittests/test-edi-energy-mirror-repo/edi_energy_de/FV2310/actual-output/COMDIS')
# ---
# name: TestCli.test_kohlrahbi_cli_with_valid_arguments[proof of concept].3
  WindowsPath('C:/GitRepos/kohlrahbi/unittests/test-edi-energy-mirror-repo/edi_energy_de/FV2310/actual-output/ORDERS')
# ---
# name: TestCli.test_kohlrahbi_cli_with_valid_arguments[proof of concept].4
  WindowsPath('C:/GitRepos/kohlrahbi/unittests/test-edi-energy-mirror-repo/edi_energy_de/FV2310/actual-output/ORDRSP')
# ---
# name: TestCli.test_kohlrahbi_cli_with_valid_arguments[proof of concept].5
  WindowsPath('C:/GitRepos/kohlrahbi/unittests/test-edi-energy-mirror-repo/edi_energy_de/FV2310/actual-output/PARTIN')
# ---
